libs/stats/odc/stats/_cli_save_tasks.py

```diff
@@ -169,26 +169,17 @@ def gqa_predicate(ds):
 def _parse_products(dc, products, temporal_range):
         
     query = {"product": products}
-    query.update(
-            temporal_range.dc_query(pad=0.6)
-        ) 
-    dss = ordered_dss(
-        dc, key=lambda ds: (ds.center_time, ds.metadata.region_code), **query
-    )
+    query.update(temporal_range.dc_query(pad=0.6)) 
+    dss = ordered_dss(dc, key=lambda ds: (ds.center_time, ds.metadata.region_code), **query)
     paired_dss = groupby(dss, key=lambda ds: (ds.center_time, ds.metadata.region_code))
-    paired_dss = [list(t[1]) for t in paired_dss]
-    bad_dss = [x for x in paired_dss if len(x) != 2]
-    paired_dss = [x for x in paired_dss if len(x) == 2]
-    
-    for x in groupby(bad_dss, key=lambda ds: ds[0].type.name):
-        name, _iter = x
-        print(f"Warning: product {name} has {len(list(_iter))} unpaired datasets")
-    
+    paired_dss = (tuple(t[1]) for t in paired_dss)
+    paired_dss = (x for x in paired_dss if len(x) == len(products))
```

Comment:
so with this logic we have no record of mismatched datasets

Reply:
We have a record of how many we expect (n_dss) and the number that are actually paired. So we have the number of datasets that are missing but no other info.

I could do something like:
- split into paired and bad dss tiers with tee
- Iterate over paired first
- Iterate over bad

But this could still yield OOM if the number of bad datasets is large. What do you think?

Reply:
I would do filtering iterator that keeps count and maybe logs things

```python
def _filter_bad_with_logging(groups, error_state_object):
    for g in groups:
       if group_no_good(g):
          error_state_object.bad_count += 1
          log_count_whatever(g, error_state_object) # update whatever counts in error_state_object and maybe log issues
       else:
          yield g
#...
errors = ... # SimpleNamescpase(...) or some dataclass or whatever
clean_stream = _filter_bad_with_logging(dirty_stream, errors)
```